wsd/RequestVettingStation.cpp

```diff
@@ -385,20 +366,51 @@ void RequestVettingStation::transferToDocBroker(const std::string& url,
 #endif
         }
     }
+    return bTransferred;
+}
+
+bool RequestVettingStation::doPostponedTransferToDocBroker(SocketDisposition& disposition)
+{
+    if (!_postponingTransferToDocBroker)
+        return false;
+
+    if (!_ws)
+    {
+        // RequestVettingStation::handleRequest hasn't been called yet. Can't
+        // transfer yet anyway.
+        return false;
+    }
+
+    // RequestVettingStation::handleRequest was called, but deferred creating
+    // the DocBroker and transferring the disposition's Socket to its
+    // SocketPoll until this result was available.
+    if (_checkFileInfo && _checkFileInfo->state() == CheckFileInfo::State::Pass &&
+        _checkFileInfo->wopiInfo())
+    {
+        _postponingTransferToDocBroker = false;
+
+        SharedSettings sharedSettings(_checkFileInfo->wopiInfo());
+        return transferToDocBroker(_checkFileInfo->url().toString(),
+                                   sharedSettings.getConfigId(),
+                                   _checkFileInfo->getSslVerifyMessage(),
+                                   _ws, disposition);
+    }
+
+    return false;
 }
 
 void RequestVettingStation::checkFileInfo(const Poco::URI& uri, int redirectLimit)
 {
-    auto cfiContinuation = [this](CheckFileInfo& checkFileInfo)
+    auto cfiContinuation = [this]([[maybe_unused]] CheckFileInfo& checkFileInfo)
     {
         assert(&checkFileInfo == _checkFileInfo.get() && "Unknown CheckFileInfo instance");
         if (_checkFileInfo && _checkFileInfo->state() == CheckFileInfo::State::Pass &&
             _checkFileInfo->wopiInfo())
         {
-            SharedSettings sharedSettings(_checkFileInfo->wopiInfo());
-            transferToDocBroker(checkFileInfo.url().toString(),
-                                sharedSettings.getConfigId(),
-                                checkFileInfo.getSslVerifyMessage());
+            // We have a useful result, wakeup SocketPoll so
+            // ClientRequestDispatcher::checkTransfer will be called
+            // promptly on the next poll.
+            _poll->wakeup();
```

Comment:
It is rather unusual that a wakeup is necessary - surely outside of the SocketPoll implementaiton, and/or for shutdown horrors (?)